Tag and check for release candidates

pat-s opened the pull request:
fixes #60 

- Creates RC tags in `release()`
- Aborts the build if  the build is running in "state = release" and the sha of the most recent commit == sha of the most recent tag && "-rc" is found in the most recent tag name
- Deletes all RC tags in `post_release()`
- Uses `gh` (if avail) to merge a PR instead of merging the branch.

## ToDo

- [x] Enumerate RC tags because we can have multiple rejections

pat-s commented:
@krlmlr Successfully used this branch to release mlr v2.18.0 including to failed attempt (which triggered two "-rc" tags).

krlmlr commented:
I'm moving towards bumping the version with each attempted release. It's also recommended practice by CRAN.

pat-s commented:
Not a friend of this practice.

Does this mean you do not want to merge this PR at all?

krlmlr commented:
Numbers are cheap.

We might support both ways, it's low priority at the moment.

pat-s commented:
I don't mind the numbers per se (I'd there is a need for a new release) but the inconsistency in news numbering (personal splean probably).

pat-s commented:
Here's a snapshot how `unleash("patch")` looks like currently (using the `tag_release_candidates_gert` branch.

<details>
<summary>pre_release()</summary>

<img width="327" alt="image" src="https://user-images.githubusercontent.com/8430564/103638321-0b901480-4f4d-11eb-9637-52e26069de9d.png">

</details>

<details>
<summary>release()</summary>

<img width="394" alt="image" src="https://user-images.githubusercontent.com/8430564/103655253-443ce780-4f67-11eb-9042-e59233347c60.png">


</details>

<details>
<summary>post_release()</summary>

<img width="277" alt="image" src="https://user-images.githubusercontent.com/8430564/103741373-3a63c480-4ff9-11eb-824a-d28550f86855.png">


</details>

krlmlr commented on R/auto.R:
```diff
@@ -325,13 +334,36 @@ post_release_impl <- function() {
+  # FIXME: clean up mess
+
+  # check if gh is avail
+  is_gh_avail <- suppressWarnings(try(system2("gh", stdout = FALSE, stderr = FALSE), silent = TRUE))
```
I'd rather discuss this change in a separate PR.

krlmlr replied:
Separate = third.

krlmlr reviewed (changes requested):
Thanks. Can you please isolate the changes so that they apply directly to the `f-27-automate` branch, and open a new PR? I'm not ready to commit to state management, RC tagging is useful in itself.

pat-s commented on R/auto.R:
```diff
@@ -64,10 +64,12 @@ pre_release_impl <- function(which, force) {
   # push main branch, bump to devel version and push again
   push_to_new(remote_name, force)
   switch_branch(main_branch)
+  push_head(main_branch)
```
not sure if these two `push_head()` calls are still needed - guess we need to check in action

pat-s commented:
Rebased and stripped the diff, not sure if I removed something important though. 

I'd just check how it behaves in practice then but ofc this is your decision in the end.

krlmlr commented:
/style

krlmlr commented:
/merge

pat-s commented:
Did your latest comment have the intention to merge this PR?

krlmlr commented:
No.

krlmlr commented:
/merge

krlmlr commented:
I'm experimenting with a flow that allows seamless integration of the target branch into the source branch. I do wonder why GitHub hasn't implemented this yet. Anyway, this is now available through the "/ merge" command in pull requests. Because GHA won't run checks when I push directly, I open a PR instead so that the user can merge manually and trigger checks in the target branch.

CC @wael-sadek.

pat-s commented:
@krlmlr How do you want to proceed here? I assume closing?

krlmlr commented:
I'm open to discussing release candidate tags. About to merge that long-running PR, this will probably close this PR.
